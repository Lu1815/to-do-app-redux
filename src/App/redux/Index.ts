export * from "./Store";
export * from "./TodosSlice";