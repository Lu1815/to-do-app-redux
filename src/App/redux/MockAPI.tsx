import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'

type MockApiDTO = {
  name: string,
  avatar: string,
  id: string
}

// Define a service using a base URL and expected endpoints
export const mockApi = createApi({
  reducerPath: 'mockApi',
  baseQuery: fetchBaseQuery({ baseUrl: 'https://6172cfe5110a740017222e2b.mockapi.io/' }),
  endpoints: (builder) => ({
    getElementsList: builder.query<MockApiDTO[], void>({
      query: () => `elements`,
    }),
  }),
})

export const { useGetElementsListQuery } = mockApi