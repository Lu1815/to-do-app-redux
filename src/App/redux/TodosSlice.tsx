import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'

export interface ITodoState {
  id: number,
  name: string,
}

const initialState: ITodoState[] = []

export const counterSlice = createSlice({
  name: 'todos-list',
  initialState,
  reducers: {
    create: (state, action: PayloadAction<ITodoState>) => {
      state.push(action.payload)
    },
  },
})

// Action creators are generated for each case reducer function
export const { create } = counterSlice.actions

export default counterSlice.reducer