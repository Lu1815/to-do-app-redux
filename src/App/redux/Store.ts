import { configureStore } from "@reduxjs/toolkit";
import todosSlice from "./TodosSlice";
import { mockApi } from "./MockAPI";

export const store = configureStore({
    reducer: {
        todos: todosSlice,
        [mockApi.reducerPath]: mockApi.reducer,
    },
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware().concat(mockApi.middleware),
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch