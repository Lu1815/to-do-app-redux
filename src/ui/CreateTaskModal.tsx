import { Dispatch, SetStateAction, useState } from "react";
import Input from "./Input";
import { useAppDispatch, useAppSelector } from "../App/redux/Hooks";
import { ITodoState, create } from "../App/redux/TodosSlice";
import { ToastContainer, toast } from "react-toastify";
import { ErrorNotification } from "../lib/Notifications";

const DEFAULT_INT = 0;

interface ICreateCardModalProps {
    showModal: boolean;
    setShowModal: Dispatch<SetStateAction<boolean>>;
}

const CreateTaskModal = ({ showModal, setShowModal }: ICreateCardModalProps) => {
    const todos: ITodoState[] = useAppSelector((state) => state.todos)
    const dispatch = useAppDispatch()
    const [taskName, setTaskName] = useState<string>("");
    const [error, setError] = useState<boolean>(false);
    const inputProps = { taskName, setTaskName, error, setError };

    const handleHideModal = () => {
        setShowModal(false)
    }

    const handleAddTask = () => {
        if (!taskName) {
            ErrorNotification('Task name cannot be empty!')
            setError(true);
            return;
        }

        let todoId: number = DEFAULT_INT;

        if(todos.length <= 0){
            todoId = 1;
        } else {
            todoId = todos[todos.length - 1].id + 1;
        }

        const todoItem: ITodoState = {
            id: todoId,
            name: taskName,
        }

        dispatch(create(todoItem))
        setTaskName("");
        setShowModal(false)
    }

    return (
        <>
            {
                showModal && (
                    <div className="bg-black bg-opacity-50 h-svh w-max">
                        <div id="popup-modal" tabIndex={-1} data-modal-backdrop="static" className="fixed inset-0 flex items-center justify-center h-screen w-screen bg-black bg-opacity-50">
                            <div className="relative p-4 w-full max-w-md max-h-full">
                                <div className="relative bg-white rounded-lg shadow">
                                    <div className="p-4 md:p-5 text-center flex flex-col">
                                        <h3 className="mb-5 text-lg font-normal text-gray-500 dark:text-gray-400">Task Name:</h3>
                                        <Input {...inputProps} />
                                        <div>
                                            <button onClick={handleHideModal} type="button" className="text-white bg-red-600 hover:bg-red-800 focus:ring-4 focus:outline-none focus:ring-red-300 dark:focus:ring-red-800 font-medium rounded-lg text-sm inline-flex items-center px-5 py-2.5 text-center me-2 mt-2">
                                                Cancel
                                            </button>
                                            <button onClick={handleAddTask} type="button" className="bg-btn hover:bg-btn-hovered focus:ring-4 focus:outline-none focus:ring-gray-200 rounded-lg border border-gray-200 text-sm font-medium px-5 py-2.5 focus:z-10">
                                                Add Task
                                            </button>

                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                )
            }
            <ToastContainer />
        </>
    )
}

export default CreateTaskModal