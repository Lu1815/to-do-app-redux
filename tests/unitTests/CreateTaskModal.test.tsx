import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'; // For additional matchers
import userEvent from '@testing-library/user-event'; // For simulating user events
import CreateTaskModal from '../../src/ui/CreateTaskModal';
import { Provider } from 'react-redux';
import { store } from '../../src/App/redux/Store';
import { delay } from 'msw';

jest.mock('../../src/App/redux/Hooks', () => ({
    useAppDispatch: jest.fn(),
    useAppSelector: jest.fn(),
}));

describe('CreateTaskModal Component', () => {
    const defaultProps = {
        showModal: true,
        setShowModal: jest.fn(),
    };

    test('renders CreateTaskModal correctly', () => {
        render(
            <Provider store={store}> {/* Wrap your component with Provider and provide the test store */}
                <CreateTaskModal {...defaultProps} />
            </Provider>
        );
        expect(screen.getByText('Task Name:')).toBeInTheDocument();
    });


    test('calls setShowModal with false when Cancel button is clicked', () => {
        render(<CreateTaskModal {...defaultProps} />);
        fireEvent.click(screen.getByText('Cancel'));
        expect(defaultProps.setShowModal).toHaveBeenCalledWith(false);
    });

    test('shows error message and does not call dispatch when Add Task button is clicked with empty taskName', async () => {
        render(
            <Provider store={store}>
                <CreateTaskModal {...defaultProps} />
            </Provider>
        );

        fireEvent.click(screen.getByText('Add Task'));

        // Wait for the asynchronous behavior (e.g., state update) to settle
        await waitFor(() => {
            expect(screen.getByText('Task name cannot be empty!')).toBeInTheDocument();
        });

        expect(defaultProps.setShowModal).toHaveBeenCalledWith(false);
    });

    // test('calls dispatch and resets state when Add Task button is clicked with a valid taskName', () => {
    //     const mockDispatch = jest.fn();
    //     const mockUseAppDispatch = jest.spyOn(require('../../src/App/redux/Hooks'), 'useAppDispatch');
    //     mockUseAppDispatch.mockReturnValue(mockDispatch);

    //     render(
    //         <Provider store={store}>
    //             <CreateTaskModal {...defaultProps} />
    //         </Provider>
    //     );

    //     userEvent.type(screen.getByLabelText('Task Name:'), 'New Task');
    //     fireEvent.click(screen.getByText('Add Task'));

    //     expect(mockDispatch).toHaveBeenCalled();
    //     expect(defaultProps.setShowModal).toHaveBeenCalledWith(false);
    //     // Additional assertions based on your implementation
    // });

    // Add more test cases as needed

    test('matches snapshot', () => {
        const { container } = render(
            <Provider store={store}> {/* Wrap your component with Provider and provide the test store */}
                <CreateTaskModal {...defaultProps} />
            </Provider>
        );
        expect(container).toMatchSnapshot();
    });
});
