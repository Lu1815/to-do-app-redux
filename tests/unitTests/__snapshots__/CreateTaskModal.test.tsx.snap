// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTaskModal Component matches snapshot 1`] = `
<div>
   
  <div
    class="bg-black bg-opacity-50 h-svh w-max"
  >
    <div
      class="fixed inset-0 flex items-center justify-center h-screen w-screen bg-black bg-opacity-50"
      data-modal-backdrop="static"
      id="popup-modal"
      tabindex="-1"
    >
      <div
        class="relative p-4 w-full max-w-md max-h-full"
      >
        <div
          class="relative bg-white rounded-lg shadow"
        >
          <div
            class="p-4 md:p-5 text-center flex flex-col"
          >
            <h3
              class="mb-5 text-lg font-normal text-gray-500 dark:text-gray-400"
            >
              Task Name:
            </h3>
            <input
              class="rounded-md p-2 border-gray-200 border-2 outline-none placeholder:text-gray-600 text-black"
              id="task-name"
              placeholder="Task Name..."
              type="text"
              value=""
            />
            <div>
              <button
                class="text-white bg-red-600 hover:bg-red-800 focus:ring-4 focus:outline-none focus:ring-red-300 dark:focus:ring-red-800 font-medium rounded-lg text-sm inline-flex items-center px-5 py-2.5 text-center me-2 mt-2"
                type="button"
              >
                Cancel
              </button>
              <button
                class="bg-btn hover:bg-btn-hovered focus:ring-4 focus:outline-none focus:ring-gray-200 rounded-lg border border-gray-200 text-sm font-medium px-5 py-2.5 focus:z-10"
                type="button"
              >
                Add Task
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="Toastify"
  />
</div>
`;
